File: exercises/exercise-2-outbox-java/src/main/java/de/emaarco/example/application/service/SendWelcomeMailService.java
package de.emaarco.example.application.service;

import de.emaarco.example.application.port.in.SendWelcomeMailUseCase;
import de.emaarco.example.application.port.out.NewsletterSubscriptionRepository;
import de.emaarco.example.application.port.out.ProcessedOperationRepository;
import de.emaarco.example.domain.OperationId;
import de.emaarco.example.domain.SubscriptionId;
import jakarta.transaction.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
@Transactional
public class SendWelcomeMailService implements SendWelcomeMailUseCase {

    private static final Logger log = LoggerFactory.getLogger(SendWelcomeMailService.class);

    private final NewsletterSubscriptionRepository repository;
    private final ProcessedOperationRepository processedOperationRepository;

    public SendWelcomeMailService(
            NewsletterSubscriptionRepository repository,
            ProcessedOperationRepository processedOperationRepository
    ) {
        this.repository = repository;
        this.processedOperationRepository = processedOperationRepository;
    }

    @Override
    public void sendWelcomeMail(SubscriptionId subscriptionId, OperationId operationId) {
        throw new UnsupportedOperationException("TODO: Implement Check-Execute-Record pattern for idempotency");
        // STEP 1: Check if already processed (return early if yes)

        // STEP 2: Execute business logic

        // STEP 3: Record operation as completed
    }
}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/io/miragon/example/application/service/newsletter/SendWelcomeMailService.kt
package io.miragon.example.application.service.newsletter

import io.miragon.example.application.port.inbound.newsletter.SendWelcomeMailUseCase
import io.miragon.example.application.port.outbound.newsletter.NewsletterSubscriptionRepository
import io.miragon.example.domain.SubscriptionId
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
class SendWelcomeMailService(
    private val repository: NewsletterSubscriptionRepository,
) : SendWelcomeMailUseCase {

    private val log = KotlinLogging.logger {}

    override fun sendWelcomeMail(subscriptionId: SubscriptionId) {
        val subscription = repository.find(subscriptionId)
        log.info { "Sending welcome mail to ${subscription.email}" }
    }
}